checker/src/org/checkerframework/checker/index/upperbound/UpperBoundTransfer.java

```diff
@@ -223,6 +253,10 @@ protected void refineGT(
             knownToBeLessThanLengthOf(array, right, store, in);
         }
 
+        if (right instanceof NumericalAdditionNode || right instanceof NumericalSubtractionNode) {
+            propagateToSubterms((BinaryOperationNode) right, refinedRight, store);
```

Comment:
Sorry for the delay!  I'll take another look at the methods.  I have an idea of how to combine them, but it will be faster for me to just try it then to explain it.